state-chain/pallets/cf-broadcast/src/migrations/v3.rs

```diff
@@ -0,0 +1,271 @@
+				broadcast_id,
+				BTreeSet::from_iter(failed_broadcasters),
+			)
+		});
+
+		let retry_queue = old::BroadcastRetryQueue::<T, I>::take()
+			.unwrap_or_default()
+			.into_iter()
+			.map(|attempt| BroadcastAttempt::<T, I> {
+				broadcast_id: attempt.broadcast_attempt_id.broadcast_id,
+				transaction_payload: attempt.transaction_payload,
+				threshold_signature_payload: attempt.threshold_signature_payload,
+				transaction_out_id: attempt.transaction_out_id,
+			})
+			.collect::<Vec<_>>();
+		BroadcastRetryQueue::<T, I>::put(retry_queue);
+
+		Weight::zero()
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, DispatchError> {
+		let broadcasts_with_failed_broadcasters =
+			old::FailedBroadcasters::<T, I>::iter_keys().collect::<Vec<_>>();
+		let broadcast_retry_queue = old::BroadcastRetryQueue::<T, I>::get()
+			.unwrap_or_default()
+			.into_iter()
+			.map(|attempt| attempt.broadcast_attempt_id.broadcast_id)
+			.collect::<Vec<_>>();
+		Ok(MigrationVerification { broadcasts_with_failed_broadcasters, broadcast_retry_queue }
+			.encode())
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade(state: Vec<u8>) -> Result<(), DispatchError> {
+		let pending_broadcasts = PendingBroadcasts::<T, I>::get();
+		for id in AwaitingBroadcast::<T, I>::iter_keys() {
+			assert!(pending_broadcasts.contains(&id));
+		}
+
+		let MigrationVerification { broadcasts_with_failed_broadcasters, broadcast_retry_queue } =
+			<MigrationVerification>::decode(&mut &state[..]).unwrap();
+
+		broadcasts_with_failed_broadcasters.into_iter().for_each(|broadcast_id| {
+			assert!(FailedBroadcasters::<T, I>::contains_key(broadcast_id))
+		});
+
+		let new_retry_queue = BroadcastRetryQueue::<T, I>::get()
+			.into_iter()
+			.map(|attempt| attempt.broadcast_id)
+			.collect::<Vec<_>>();
+
+		broadcast_retry_queue
+			.into_iter()
+			.for_each(|broadcast_id| assert!(new_retry_queue.contains(&broadcast_id)));
+
+		Ok(())
+	}
+}
+
+#[cfg(test)]
+mod migration_tests {
+	use super::*;
+	use crate::mock::*;
+
+	#[test]
+	fn migration_works_v2_to_v3_awaiting_broadcast() {
+		new_test_ext().execute_with(|| {
+			// Insert mock data into old storage
+			for broadcast_id in 1..10 {
+				for attempt_id in 0..=broadcast_id {
+					let broadcast_attempt_id =
+						old::BroadcastAttemptId { broadcast_id, attempt_count: attempt_id };
+					let singing_attempt = old::TransactionSigningAttempt {
+						broadcast_attempt: old::BroadcastAttempt {
+							broadcast_attempt_id,
+							transaction_payload: Default::default(),
+							threshold_signature_payload: Default::default(),
+							transaction_out_id: Default::default(),
+						},
+						nominee: attempt_id as u64,
+					};
+					old::AwaitingBroadcast::<Test, Instance1>::insert(
+						broadcast_attempt_id,
+						singing_attempt,
+					);
+				}
+			}
+
+			// Perform runtime migration.
+			crate::migrations::v3::Migration::<Test, Instance1>::on_runtime_upgrade();
```

Comment:
You can also run the pre-and post upgrade hooks in the test, before and after the upgrade respectively - just have to make the tests with the `try-runtime` flag too. I think that's worth doing